### PR TITLE
Added advanced push samples

### PR DESCRIPTION
**Pull request by Sergio-Mira**
https://swrvedev.jira.com/browse/SWRVE-12394

### CODE REVIEW

File: samples/AdvancedPushNotifications/README [hunk 1]
```diff
@@ -0,0 +1,14 @@
+Eclipse ADT build instructions
+==================================================
+-Include SwrveSDK project in workspace.
+-Include SwrveSDKDemo project in workspace.
+-Replace YOUR_APP_ID in MainActivity.java with your app id.
+-Replace YOUR_API_KEY in MainActivity.java with your api key.
```
Comment (dominicmarmionswrve) [inline samples/AdvancedPushNotifications/README]: In DemoApplication, not MainActivity

File: samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/BigNotificationSwrveGcmIntentService.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+            // Set the large icon downloaded from the custom payload 'image_url' or a default
+            Bitmap largeBitmap = BitmapFactory.decodeResource(getResources(), R.drawable.default_large_image);
+            String imageUrl = msg.getString("image_url");
+            if (!SwrveHelper.isNullOrEmpty(imageUrl)) {
+                try {
+                    largeBitmap = Picasso.with(this).load(imageUrl).get();
+                } catch (IOException e) {
+                    // Download failed
+                    e.printStackTrace();
```
Comment (dominicmarmionswrve) [inline samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/BigNotificationSwrveGcmIntentService.java]: Use either the SwrveLogger or the normal android Log.e

File: samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/DemoApplication.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+import android.app.Application;
+import android.os.Bundle;
+import android.util.Log;
+
+import com.swrve.sdk.SwrveSDK;
+import com.swrve.sdk.config.SwrveConfig;
+import com.swrve.sdk.gcm.ISwrvePushNotificationListener;
+
+public class DemoApplication extends Application {
```
Comment (dominicmarmionswrve) [inline samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/DemoApplication.java]: Rename this application class. Its not the demo.

File: samples/AdvancedPushNotifications/README (hunk 1 above)
Comment (dominicmarmionswrve) [inline samples/AdvancedPushNotifications/README]: Actually, rename that DemoApplication also.

File: samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/ConfigContentFragment.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    @Override
+    public void onResume() {
+        super.onResume();
+
+        try {
+            String eventsUrl = SwrveSDK.getConfig().getEventsUrl().toURI().toString();
+            TextView textViewEventsUrl = (TextView) getView().findViewById(R.id.config_eventsUrl);
+            textViewEventsUrl.setText(eventsUrl);
+        } catch (URISyntaxException ex) {
+            Log.e("Demo", "Error getting events url", ex);
```
Comment (dominicmarmionswrve) [inline samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/ConfigContentFragment.java]: Demo?

File: samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/CustomLayoutSwrveGcmIntentService.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+import com.swrve.sdk.SwrveHelper;
+import com.swrve.sdk.gcm.SwrveGcmIntentService;
+
+/*
+ * Use this class when you want to create richer local notifications form a push notification
+ * using a custom layout.
+ * http://developer.android.com/guide/topics/ui/notifiers/notifications.html
+ */
+public class CustomLayoutSwrveGcmIntentService extends SwrveGcmIntentService {
```
Comment (dominicmarmionswrve) [inline samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/CustomLayoutSwrveGcmIntentService.java]: Not sure we need this? The BigNotificationSwrveGcmIntentService already shows how to override the createNotification. That should be enough to demonstrate people can override this method and return whatever type of notification they like.

File: samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/DemoApplication.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+            config.setSenderId(YOUR_SENDER_ID);
+            SwrveSDK.createInstance(this, YOUR_APP_ID, YOUR_API_KEY, config);
+            // React to the push notification when the user clicks on it
+            SwrveSDK.setPushNotificationListener(new ISwrvePushNotificationListener() {
+                @Override
+                public void onPushNotification(Bundle bundle) {
+                    if (bundle.containsKey("custom_key")) {
+                        String customValue = bundle.getString("custom_key");
+                        // Do something awesome with custom value!
+                        Log.d("DemoApplication", "Received push payload " + customValue);
```
Comment (dominicmarmionswrve) [inline samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/DemoApplication.java]: DemoApplication

File: samples/AdvancedPushNotifications/README [hunk 7]
```diff
@@ -0,0 +1,14 @@
+Eclipse ADT build instructions
```
Comment (dominicmarmionswrve) [inline samples/AdvancedPushNotifications/README]: Eclipse instructions? Don't think we need this or should support this.

File: samples/AdvancedPushNotifications/README [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+-Replace YOUR_APP_ID in MainActivity.java with your app id.
+-Replace YOUR_API_KEY in MainActivity.java with your api key.
+-Run SwrveSDKDemo app normally.
+
+Android Studio build instructions
+==================================================
+-Import SwrveSDKDemo.
+-Replace YOUR_APP_ID in MainActivity.java with your app id.
+-Replace YOUR_API_KEY in MainActivity.java with your api key.
+-Run SwrveSDKDemo app normally.
```
Comment (dominicmarmionswrve) [inline samples/AdvancedPushNotifications/README]: Can you add more details to the readme describing what the the sample is demonstrating? And where the important pieces are.

File: samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/BigNotificationSwrveGcmIntentService.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+            if (SwrveHelper.isNullOrEmpty("group")) {
+                msgGroup = "group_key_emails";
+            }
+
+            // Set the large icon downloaded from the custom payload 'image_url' or a default
+            Bitmap largeBitmap = BitmapFactory.decodeResource(getResources(), R.drawable.default_large_image);
+            String imageUrl = msg.getString("image_url");
+            if (!SwrveHelper.isNullOrEmpty(imageUrl)) {
+                try {
+                    largeBitmap = Picasso.with(this).load(imageUrl).get();
```
Comment (dominicmarmionswrve) [inline samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/BigNotificationSwrveGcmIntentService.java]: Can we not use Picasso unless we have a good use case for it? I don't think its needed for this use case.

File: samples/IntegrateWithOtherPushProvider/README [hunk 10]
```diff
@@ -0,0 +1,14 @@
+Eclipse ADT build instructions
+==================================================
+-Include SwrveSDK project in workspace.
+-Include SwrveSDKDemo project in workspace.
+-Replace YOUR_APP_ID in MainActivity.java with your app id.
+-Replace YOUR_API_KEY in MainActivity.java with your api key.
+-Run SwrveSDKDemo app normally.
+
+Android Studio build instructions
+==================================================
```
Comment (dominicmarmionswrve) [inline samples/IntegrateWithOtherPushProvider/README]: Same as other Readme. Remove eclipse instructions, give more details for what is actually being demo'ed. Also can we rename the project? Its a bit wordy. Suggestion: MultiGcm or MultiplePushProviders??

File: samples/IntegrateWithOtherPushProvider/src/main/java/com/swrve/sdk/demo/ConfigContentFragment.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    @Override
+    public void onResume() {
+        super.onResume();
+
+        try {
+            String eventsUrl = SwrveSDK.getConfig().getEventsUrl().toURI().toString();
+            TextView textViewEventsUrl = (TextView) getView().findViewById(R.id.config_eventsUrl);
+            textViewEventsUrl.setText(eventsUrl);
+        } catch (URISyntaxException ex) {
+            Log.e("Demo", "Error getting events url", ex);
```
Comment (dominicmarmionswrve) [inline samples/IntegrateWithOtherPushProvider/src/main/java/com/swrve/sdk/demo/ConfigContentFragment.java]: Demo?

File: samples/IntegrateWithOtherPushProvider/src/main/java/com/swrve/sdk/demo/ConfigContentFragment.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.TextView;
+
+import com.swrve.sdk.SwrveSDK;
+
+import java.net.URISyntaxException;
+
+public class ConfigContentFragment extends Fragment {
```
Comment (dominicmarmionswrve) [inline samples/IntegrateWithOtherPushProvider/src/main/java/com/swrve/sdk/demo/ConfigContentFragment.java]: I would remove this class completely and just have a MainActivity (same for the AdvancedPushNotifications project). There should be no UI in the app. The MainActivity should have text to describe what this sample is doing.

File: samples/IntegrateWithOtherPushProvider/src/main/java/com/swrve/sdk/demo/MainActivity.java [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+    private View root;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        SwrveSDK.onCreate(this);
+
+        root = getLayoutInflater().inflate(R.layout.activity_main, null);
```
Comment (dominicmarmionswrve) [inline samples/IntegrateWithOtherPushProvider/src/main/java/com/swrve/sdk/demo/MainActivity.java]: I'd cut this class down to the bare minimum.......No need for UI, toolbar, etc.

File: samples/MessageCenter/README [hunk 14]
```diff
@@ -0,0 +1,14 @@
+Eclipse ADT build instructions
+==================================================
+-Include SwrveSDK project in workspace.
+-Include SwrveSDKDemo project in workspace.
+-Replace YOUR_APP_ID in MainActivity.java with your app id.
+-Replace YOUR_API_KEY in MainActivity.java with your api key.
+-Run SwrveSDKDemo app normally.
+
+Android Studio build instructions
```
Comment (dominicmarmionswrve) [inline samples/MessageCenter/README]: Same comments as other readme's

File: samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/BigNotificationSwrveGcmIntentService.java (hunk 9 above)
Comment (Sergio-Mira) [inline samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/BigNotificationSwrveGcmIntentService.java]: Provides caching etc. I think it is good to not have to reimplement, or have a sample that is not complete (bad image downloads)

Comment (dominicmarmionswrve) [inline samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/BigNotificationSwrveGcmIntentService.java]: Why bother downloading the image? The sample should just demo how to create a custom notification. 

I love Picasso, its a great library, but don't want to pull it into our code without a good reason. This isn't a good enough reason.

File: samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/DemoApplication.java (hunk 6 above)
Comment (Sergio-Mira) [inline samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/DemoApplication.java]: Not sure what you mean here

File: samples/AdvancedPushNotifications/README (hunk 7 above)
Comment (Sergio-Mira) [inline samples/AdvancedPushNotifications/README]: They are all copies from the main demo that were not updated, so nothing new. Will remove from them all

File: samples/IntegrateWithOtherPushProvider/README (hunk 10 above)
Comment (Sergio-Mira) [inline samples/IntegrateWithOtherPushProvider/README]: Yeah MultiplePushProviders sounds good

File: samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/BigNotificationSwrveGcmIntentService.java (hunk 9 above)
Comment (Sergio-Mira) [inline samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/BigNotificationSwrveGcmIntentService.java]: We are not pulling it into our SDK, just the samples. And a customer use case is specifically downloading an image from a custom payload URL.

The more in depth examples we give the less work they and CSM have to do.

Comment (dominicmarmionswrve) [inline samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/BigNotificationSwrveGcmIntentService.java]: Yeah, I know its not in the sdk we give customers, its in the samples. But why do we need to put it in the attributions? 

Plus someone could start using it in a PR in the main sdk, thinking its ok to bring it in there because its in the samples.

Its not needed for the use case.

File: samples/AdvancedPushNotifications/README (hunk 7 above)
Comment (dominicmarmionswrve) [inline samples/AdvancedPushNotifications/README]: 👍

File: samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/DemoApplication.java (hunk 6 above)
Comment (dominicmarmionswrve) [inline samples/AdvancedPushNotifications/src/main/java/com/swrve/sdk/demo/DemoApplication.java]: Same point I guess as the others. Should this actually be called DemoApplication? Plus there's already a LOG_TAG defined in the class, so use that.....

**Comment by dominicmarmionswrve** [comment]
Great readme's! And the samples are good examples of how to use Swrve! 🚀

**Comment by dominicmarmionswrve** [comment]
🚀
